nixos/tests/systemd-confinement.nix

```diff
@@ -80,6 +86,77 @@ import ./make-test-python.nix {
               machine.fail("chroot-exec touch /bin/test")
         '';
       }
+      { config.confinement.mode = "full-apivfs";
+        config.serviceConfig.DynamicUser = true;
+        testScript = ''
+          with subtest("check if DynamicUser is working in full-apivfs mode"):
+              machine.succeed("chroot-exec ls -l /dev")
+              paths = machine.succeed('chroot-exec find / -path /dev/"\\*" -prune -o -path /nix/"\\*" -prune -o -path /proc/"\\*" -prune -o -path /sys/"\\*" -prune -o -print || test $? = 1')
+              assert_eq(
+                '\n'.join(sorted(paths.split('\n'))),
+          """
+          /
+          /bin
+          /bin/sh
+          /dev
+          /etc
+          /nix
+          /proc
+          /root
+          /run
+          /run/host
+          /run/host/.os-release-stage
+          /run/host/.os-release-stage/os-release
+          /run/host/os-release
+          /run/systemd
+          /run/systemd/incoming
+          /sys
+          /tmp
+          /usr
+          /var
+          /var/tmp
+          find: '/root': Permission denied
+          find: '/run/systemd/incoming': Permission denied"""
+              )
+              uid = machine.succeed('chroot-exec id -u').strip()
+              assert uid != "0", "UID of a DynamicUser shouldn't be 0"
+              machine.fail("chroot-exec touch /bin/test")
+              # DynamicUser=true implies ProtectSystem=strict
+              machine.fail("chroot-exec touch /etc/test")
+        '';
+      }
+      { config.confinement.mode = "chroot-only";
+        config.serviceConfig.DynamicUser = true;
+        testScript = ''
+          with subtest("check if DynamicUser is working in chroot-only mode"):
+              paths = machine.succeed('chroot-exec find / -path /nix/"\\*" -prune -o -print || test $? = 1')
+              assert_eq(
+                '\n'.join(sorted(paths.split('\n'))),
+          """
+          /
+          /bin
+          /bin/sh
+          /dev
+          /etc
+          /nix
+          /proc
+          /root
+          /run
+          /run/systemd
+          /run/systemd/incoming
+          /sys
+          /tmp
+          /usr
+          /var
+          /var/tmp
```

Comment:
This should not be there, nor should it be writable.

Reply:
This is enforced by systemd [when `DynamicUser=true`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=):
> If DynamicUser= is enabled, RemoveIPC= and PrivateTmp= are implied (and cannot be turned off).

If need be it can still be disabled with:
```
InaccessiblePaths= [ "-+/tmp" "-+/var/tmp" ];
```
Should I add an option to control that?

Reply:
> Should I add an option to control that?

Hm, difficult to say since then explicitly providing `PrivateTmp` would then result in a no-op. On the other hand when using `systemd-confinement` with `chroot-only` mode, I pretty much expect that there is not a single directory that is writable. For example I usually start with `{ binSh = null; mode = "chroot-only"; }` as the most restrictive environment and then only lift those restrictions when absolutely necessary.

Having a single writable directory there would already implicitly lift those restrictions and the documentation you linked about why `PrivateTmp` are forced-on by default only make sense if you assume that `/tmp` and `/var/tmp` are *already there* which in this case they are not.

Adding stuff to `InaccessiblePaths` also makes it confusing, since we deviate from the systemd documentation. Additionally, "emulating" `PrivateTmp` on our side would introduce unnecessary complexity.

So maybe we should instead document on our side that if you want the *most* restrictive environment (from a filesystem perspective) possible, you have to use dedicated users.

> Should I add an option to control that?

Good question... I usually try to only introduce options whenever really necessary and in this case people can just add `InaccessiblePaths`, but given that this makes it even more verbose to start with the most restrictive configuration, a shortcut could make sense. Additionally, having such a shortcut would also mean that we can explicitly test this scenario.

Better ideas? Also, cc @martinetd on this.

Reply:
Hmm, I'm also of the opinion that chroot-only shouldn't have any writable directory by default.

OTOH, adding tmp paths to InaccessiblePaths without a way to unset it is a bit annoying -- that can be overridden with mkForce on the service's InaccessiblePaths but it's not practical to just remove a single element... Which leaves the hard part of "how".

Thinking out loud:
 - plain new option; I don't really like that as it increases the confinement API which is hard to grasp enough as it is
 - adding new modes e.g. chroot-only/apivfs/tmp; I'm not sure about that either (would it base off apivfs or chroot-only? and it's not really logical)
 - we could look if the user explicitly set ServiceConfig.PrivateTmp and add /tmp+/var/tmp to inaccesiblepaths if and only if PrivateTmp isn't true:
    - That basically overrides the default that DynamicUser implies PrivateTmp, but the rationale in the man page is that it doesn't want the service to leave files behind, which is still respected
    - A user that wants either tmp directory can explicitly request privatetmp and eventually mask the other path if they want to
    - Downside is that it's not as obvious/discoverable as a brand new option
 - Open an issue @ systemd upstream to make PrivateTmp not create the directories if they don't exist
   - That's probably the cleanest way forward: A user can somehow create the directories if they want them
   - OTOH I think it'll be a tough sell to systemd devs, and will take time to be available for us.

I'm honestly not sure what's best here... Kludging over ServiceConfig.PrivateTmp is probably a good compromise even if it's a bit ugly?

Reply:
> * e could look if the user explicitly set ServiceConfig.PrivateTmp and add /tmp+/var/tmp to inaccesiblepaths if and only if PrivateTmp isn't true:

One issue with this is that if you use eg. `systemd.packages`, it could include `PrivateTmp` and we have no way to reliably detect this at eval-time (well, except IFD of course, but...).

Reply:
I explored option 3. (piggybacking on `PrivateTmp=`), tell me if it's worth pursuing/keeping, at least until option 4. (modifying `systemd`) becomes a reality.

Reply:
Note that I had to set `InaccessiblePaths`= through `config.systemd.packages` because `serviceConfig` options cannot depend on each others without causing some obscure infinite recursion.

Reply:
I'm not fond of requiring users to explicitly pull in confinement.dynamicUser; my interest in this PR lies in that existing services become much easier to use as is now so adding more knobs somewhat murks the point. I agree an assertion + separate option makes is stand out more and will lead to less unexpected behaviors, but I expect confinement to give no writable directory more than I expect a /tmp to be available, so I think I would just find this annoying in practice.

As for the default of PrivateTmp "becoming" false I don't think that's much of a problem -- at very least:
- this won't break anything, because this only impacts services with dynamicuser enabled (I haven't found any other setting that implies it), and it's not possible to use it now, so no use case is broken
- when enabling it the user will test the service and I think it'll be obvious that /tmp is missing from any error that pops up (I don't consider confinement to be something that can be enabled and trusted to work without testing); and I find the documentation julm added to be sufficient in understanding what happened (but then again, I was here in this PR, so not in a position to judge how surprising that really is)

So I'm in favor of the current solution, but we can keep discussing if you really prefer having a separate option.

Reply:
Hm, searching on GitHub for [chroot-only language:Nix](https://github.com/search?q=chroot-only+language%3ANix&type=code) turns out that the chroot-only mode seems to be only used by a handful users (at least those making their source code available), so I guess the impact could be rather low, since switching those services to `DynamicUser` is also a conscious decision.

Reply:
(note my /etc/nixos repository is [public](https://cgit.notk.org/asmadeus/nixos-config.git/), but not on github -- and that's also true for julm (not giving a link without his permission)... I'm afraid the subset of folks using chroot-only might also be biased towards not publishing "private-ish" stuff on this platform :/ But yes, the switch back to DynamicUser is a conscious decision)